### PR TITLE
modified URI function to accept urls with mising trailing slash

### PR DESCRIPTION
**Pull request by Manvel**
Hi Guys,
I've modified the URI function so the URLs with missing trailing slash will be return host in extractHostFromURL() function:
The script is also checking whether in case of missing trailing slash is there Query sign or Fragment sign in the URL so current URLs will also works:
- http://example.com#something
- http://example.com?something
  Hope I understood the problem correctly.

### CODE REVIEW

**Comment by Manvel** [comment]
Little changed the code to make it more compact.

**Comment by palant** [comment]
This pull request should also update the unit tests in qunit/tests/baseDomain.js - the new edge cases introduced here need testing.

**Comment by Manvel** [comment]
Hi @palant 
Updated the code,
Also updated unit test, 
Please let me know if still something missing.
Didn't noticed unit test before.
Thanks for your notes.

**Comment by Manvel** [comment]
@palant Committed last changes,
In case if something is missing will update in the evening, 
Have to run now,
Thanks for your notes.

**Comment by Manvel** [comment]
@palant updated.

File: lib/basedomain.js [hunk 1]
```diff
@@ -174,9 +174,25 @@ function URI(/**String*/ spec)
...
   this._hostPortEnd = spec.indexOf("/", this._hostPortStart);
-  if (this._hostPortEnd < 0)
-    throw new Error("Invalid URI host");
+  if (this._hostPortEnd < 0) 
+  {
+    this._hostPortEnd = spec.length;
+    var queryChar = spec.indexOf("?", this._hostPortStart);
+    var fragmentChar = spec.indexOf("#", this._hostPortStart);
+    if (queryChar >= 0 && fragmentChar >= 0)
+      this._hostPortEnd = (queryChar < fragmentChar) ? queryChar : fragmentChar;
```
Comment (palant) [inline lib/basedomain.js]: Feel free to use `Math.min()` here.

File: lib/basedomain.js [hunk 2]
```diff
@@ -174,9 +174,25 @@ function URI(/**String*/ spec)
...
+    else
+    {
+       if(queryChar >= 0)
+         this._hostPortEnd = queryChar;
+       if(fragmentChar >= 0)
```
Comment (palant) [inline lib/basedomain.js]: This check is pointless if `queryChar >= 0` - you already know that `fragmentChar` has to be negative in that scenario. So you should use `else if` here. In fact, the big `else` here block is pointless - it should be `else if` as well:

```
if (queryChar >= 0 && fragmentChar >= 0)
  ...
else if (queryChar >= 0)
  ...
else if (fragmentChar >= 0)
  ...
else
  this._hostPortEnd = spec.length;
```

Note that this also takes care of pointlessly assigning `this._hostPortEnd` to `spec.length` in cases where it should have a different value.

Nits: please put a space after `if`. Also, `queryIndex` and `fragmentIndex` would describe the value of these variables better.

File: qunit/tests/baseDomain.js [hunk 3]
```diff
@@ -59,6 +64,33 @@
...
+      ["http://example.com", {
+        scheme: "http",
+        host: "example.com",
+        asciiHost: "example.com",
+        hostPort: "example.com",
+        port: -1,
+        path: "",
+        prePath: "http://example.com"
+      }],
+      ["http://example.com?", {
```
Comment (palant) [inline qunit/tests/baseDomain.js]: Test `http://example.com?foo#bar` instead to make sure `path` is set correctly in this scenario?

File: qunit/tests/baseDomain.js [hunk 4]
```diff
@@ -59,6 +64,33 @@
...
+        scheme: "http",
+        host: "example.com",
+        asciiHost: "example.com",
+        hostPort: "example.com",
+        port: -1,
+        path: "?",
+        prePath: "http://example.com"
+      }],
+      ["http://example.com#", {
```
Comment (palant) [inline qunit/tests/baseDomain.js]: Test `http://example.com#foo?bar` instead to make sure `path` is set correctly in this scenario?

**Comment by Manvel** [comment]
@palant Updated.
Thanks for all your notes.
I had to notice that.
Hope now the branch will be Okey to merge.

**Comment by palant** [comment]
This looks good enough to be merged (renaming `queryChar` into `queryIndex` and `fragmentChar` into `fragmentIndex` hasn't been done but that's not essential). Could you sign the [contributor agreement](https://adblockplus.org/eyeo-contributor-license-agreement.pdf), scan it or take a picture of it and send it to me or Felix?

**Comment by Manvel** [comment]
@palant just sent you the agreement, also changed queryChar to queryIndex and fragmentChar to fragmentIndex.
Thanks.

**Comment by palant** [comment]
Merged:
https://hg.adblockplus.org/adblockpluschrome/rev/1f3fdd417cdd
https://hg.adblockplus.org/libadblockplus/rev/f28cfdbc0032

This change should pop up on GitHub as well shortly.
